replace_with/src/lib.rs

```diff
@@ -0,0 +1,7 @@
+use std::ptr;
+
+/// Replaces `*t` with `f` applied to the original `*t`,
+pub fn replace_with<T, F: FnOnce(T) -> T>(t: &mut T, f: F) {
+    let pointer = t as *mut T;
+    unsafe { ptr::write(pointer, f(ptr::read(pointer))) }
```

Comment:
Slick! Ross discovered that you can actually use `t` in place of `pointer` in both these calls because of call-site coercion (`&mut` can coerce to both `*const` and `*mut`).

I'm not recommending this though.